src/lib/pubkey/dilithium/dilithium_common/dilithium_algos.cpp

```diff
@@ -552,17 +565,20 @@ void sample_ntt_uniform(StrongSpan<const DilithiumSeedRho> rho,
     * A generator that returns the next coefficient sampled from the XOF,
     * according to: NIST FIPS 204, Algorithm 14 (CoeffFromThreeBytes).
     */
-   auto next_coeff = [](Botan::XOF& xof) -> uint32_t {
+   auto next_coeff = [n = uint16_t(0)](Botan::XOF& xof) mutable -> uint32_t {
       std::array<uint8_t, 4> bytes = {0};
       std::span<uint8_t, 3> sampling_sink_in_bytes = std::span{bytes}.first<3>();
 
-      while(true) {
+      while(n < DilithiumConstants::SAMPLE_NTT_POLY_FROM_XOF_BOUND) {
+         n += static_cast<uint16_t>(sampling_sink_in_bytes.size());
          xof.output(sampling_sink_in_bytes);
          const auto z = load_le(bytes) & 0x7FFFFF;
          if(z < DilithiumConstants::Q) {
             return z;
          }
       }
```

Comment:
Alternatively, we could wrap the XOF in some object that counts its output bytes and throws an internal error if the bytes are exhausted. Something like:
```cpp
auto xof = LimitedXOF(mode.symmetric_primitives().H(seed), DilithiumConstants::SAMPLE_NTT_POLY_FROM_XOF_BOUND);
```
Disadvantage: This would be a complex class that must forward all necessary XOF functions. It also hides the loop count logic (which makes following the spec somewhat harder).

Advantage: No statehandling in the loop required. Also, we could combine this with the optimization, where we get all XOF bytes beforehand. Especially since we know the number XOF output bytes at compile time, we can use stack memory for this. 

I'm not sure if it's worth it, though. Just collecting ideas ;)

Reply:
Mhm, I was too lazy to do that yesterday. I'll give it a shot.

Reply:
See here: https://github.com/randombit/botan/pull/4270/commits/e529e081ce2c3a1b4b1b04bb7e0dc62332b2efd7